pvlib/ivtools.py

```diff
@@ -0,0 +1,215 @@
+
+    Returns
+    -------
+    tuple of the following elements:
+
+        photocurrent : float
+            photocurrent, A
+
+        saturation_current : float
+            dark (saturation) current, A
+
+        resistance_shunt : float
+            shunt (parallel) resistance, ohm
+
+        resistance_series : float
+            series resistance, ohm
+
+        nNsVth : float
+            product of diode (ideality) factor n (unitless) x number of
+            cells in series Ns (unitless) x cell thermal voltage Vth (V), V
+
+    References
+    ----------
+    [1] C. B. Jones, C. W. Hansen, Single Diode Parameter Extraction from
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+    # Find intercept and slope of linear portion of IV curve.
+    # Start with V < vlim * v_oc, extend by adding points until slope is
+    # acceptable
+    beta = [np.nan for i in range(5)]
+    # get index of largest voltage less than/equal to limit
+    idx = _max_index(v, vlim * v_oc)
+    while np.isnan(beta[1]) and (idx <= len(v)):
+        try:
+            coef = np.polyfit(v[:idx], i[:idx], deg=1)
+            if coef[0] < 0:
+                # intercept term
+                beta[0] = coef[1].item()
+                # sign change of slope to get positive parameter value
+                beta[1] = -coef[0].item()
+        except Exception as e:
+            raise e
+        if np.isnan(beta[1]):
+            idx += 1
+
+    if not np.isnan(beta[0]):
+        # Find parameters from exponential portion of IV curve
+        y = beta[0] - beta[1] * v - i
+        x = np.array([np.ones_like(v), v, i]).T
+        idx = _min_index(y, ilim * i_sc)
+        try:
+            result = np.linalg.lstsq(x[idx:, ], np.log(y[idx:]))
+            coef = result[0]
+            beta[3] = coef[1].item()
+            beta[4] = coef[2].item()
+        except Exception as e:
+            raise e
+
+    if not any([np.isnan(beta[i]) for i in [0, 1, 3, 4]]):
+        # calculate parameters
+        nNsVth = 1.0 / beta[3]
+        Rs = beta[4] / beta[3]
+        Gp = beta[1] / (1.0 - Rs * beta[1])
+        Rsh = 1.0 / Gp
+        IL = (1 + Gp * Rs) * beta[0]
+        # calculate I0
+        I0_v_mp = _calc_I0(IL, i_mp, v_mp, Gp, Rs, beta[3])
+        I0_v_oc = _calc_I0(IL, 0, v_oc, Gp, Rs, beta[3])
+        if (I0_v_mp > 0) and (I0_v_oc > 0):
+            I0 = 0.5 * (I0_v_mp + I0_v_oc)
+        elif (I0_v_mp > 0):
+            I0 = I0_v_mp
+        elif (I0_v_oc > 0):
+            I0 = I0_v_oc
+        else:
+            I0 = np.nan
+    else:
+        IL = I0 = Rs = Rsh = nNsVth = np.nan
+
+    return IL, I0, Rsh, Rs, nNsVth
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
+
+
+def _max_index(x, xlim):
+    """ Finds maximum index of value of x <= xlim """
+    return int(np.argwhere(x <= xlim)[-1])
```

Comment:
If I understand correctly, you're assuming that the input is sorted. If so, I think `np.searchsorted` does the job for you.

Reply:
Yes.  Thanks!